api-server/src/server/boot/challenge.js

```diff
@@ -82,6 +81,277 @@ export default async function bootChallenge(app, done) {
+  return Observable.fromPromise(updatePromise)
+    .map(() => {
+      return res.json({
+        completedChallenges,
+        points
+      });
+    })
+    .subscribe(() => {}, next);
+}
+
+/**
+ * Handles submissions for all certification projects.
+ * @param {Request} req
+ * @param {Response} res
+ * @returns {Observable}
+ */
+export function projectCompleted(req, res, next) {
+  const { user, body } = req;
+
+  // Ensure `body` is an object
+  if (!body || typeof body !== 'object') {
+    return res.status(400).json({
+      type: 'error',
+      message: 'Invalid request format. Expected `body` to be an object.'
+    });
+  }
+
+  const { id } = body;
+
+  // Find the structure, based on the `id` of the project
+  const isProjectValid = validateProject(id, body);
+
+  if (!isProjectValid) {
+    return res.status(403).json({
+      type: 'error',
+      message: 'Invalid project submission.'
+    });
+  }
+
+  const { alreadyCompleted, savedChallenges, updateData } = buildUserUpdate(
+    user,
+    id,
+    body
+  );
+
+  const points = alreadyCompleted ? user.points : user.points + 1;
+  const updatePromise = new Promise((resolve, reject) =>
+    user.updateAttributes(updateData, (err, instance) => {
+      if (err) {
+        return reject(err);
+      }
+      return resolve(instance);
+    })
+  );
+  return Observable.fromPromise(updatePromise)
+    .map(() => {
+      return res.json({
+        completedChallenges: updateData.$push.completedChallenges,
+        points,
+        savedChallenges
+      });
+    })
+    .subscribe(() => {}, next);
+}
+
+export const expectedProjectStructures = {
+  [challengeTypes.multifileCertProject]: [
+    // 14 'responsive-web-design'
+    'challengeType',
+    'files',
+    'files.contents',
+    'files.ext',
+    'files.history',
+    'files.key',
+    'files.name',
+    'id'
+  ],
+  [challengeTypes.bonfire]: [
+    // 5 'javascript-algorithms-and-data-structures'
+    'challengeType',
+    'files',
+    'files.contents',
+    'files.ext',
+    'files.history',
+    'files.key',
+    'files.name',
+    'id'
+  ],
+  [challengeTypes.zipline]: ['challengeType', 'id', 'solution'], // 3 'front-end-development-libraries'
+  [challengeTypes.backEndProject]: [
+    'challengeType',
+    'id',
+    'solution',
+    'githubLink'
+  ], // 4 'back-end-development'
+  [challengeTypes.pythonProject]: [
+    // 10 'scientific-computing-with-python'
+    'challengeType',
+    'id',
+    'solution',
+    'githubLink'
+  ],
+  [challengeTypes.codeAllyCert]: ['challengeType', 'id', 'solution'] // 13 'relational-database'
+};
+
+/**
+ * Validates the `id` submitted is a valid challenge.
+ * @param {string} id
+ * @param {number} challengeType
+ * @returns {boolean}
+ */
+export function validateChallenge(id, challengeType) {
+  // TODO: Probably can save some computation, by not calling `getChallenges` every time.
+  const challenge = getChallenges().find(chal => chal.id === id);
```

Comment:
`getChallenges` isn't desperately expensive (~0.5ms on machine), but we don't need to compute it in response to a request.  It would be better to call it once and use the result.